### ISSUE TITLE
mix plugin

### ISSUE THREAD

**Issue by ostera** [AbstractMachinesLab/caramel#30]
Right now to be able to use `caramelc` in an Erlang or Elixir project, we sort of have to manually invoke it in the right place and then also compile the caramel stdlib erlang sources and make sure its available. An entire dance.

It'd be useful to have a `mix` plugin to see what the interop from the Elixir side feels like so that we can just create a new mix project, and if you have caramel available in your path then it will do the build.

For some inspiration have a look at https://github.com/seanhinde/rebar3_caramel and #22 

The end result would be that you can drop in a couple of .ml files in your sources and have it available in the repl.

**Comment by jrusso1020** [comment]
@ostera I gave this a shot, you can see my repo here https://github.com/jrusso1020/mix_caramel let me know what you think. I haven't pushed it up to hex yet but planned to. The simple basic_project example seems to work but not sure if there's more that needs to be done.

The rebar plugin example definitely is doing a lot more

**Comment by ostera** [comment]
Hej @jrusso1020! Thanks for this 🙌🏼 -- it looks really good! 😄, feel free to submit some of your changes to the one we have hosted in here btw: https://github.com/AbstractMachinesLab/mix_caramel so we eventually converge into a single one 😃 

Re: how Gleam does it -- Gleam is partially [implementing a build system](https://github.com/AbstractMachinesLab/crane/issues/2#issue-711890213), which I won't do in Caramel. Instead, I'm building [Crane](https://github.com/AbstractMachinesLab/crane) ([and here's why](https://github.com/AbstractMachinesLab/crane/tree/main/docs)).

The reason for this is that I want to make sure that a `caramelc` can be used in other build systems too, like `dune` or `bazel`, and this is not something that can be achieved by hiding the compilation process. So `caramelc` will continue to only ask for input files and just compile them. 

Re: making outputs relocatable, this would be nice but it is definitely not a priority as it can be worked around by the build systems and at this time would mean heavier refactoring on the ocaml compiler libraries.

**Comment by jrusso1020** [comment]
Oh sorry about that, didn't realize there was already a repo 😬 . I renamed my repo, and have opened a PR for my changes. https://github.com/AbstractMachinesLab/mix_caramel/pull/1 which we can discuss there further, I'm open to keeping more of what was already there or adapting slightly.

One thing we may need to figure out is that I pushed my project up to hex already, but I think we can just overwrite it. I may need to give you all access to the project on hex though
